### ISSUE TITLE
[Performance] Add option for parser to compute Node.loc eagerly during parsing

### ISSUE THREAD

**Issue by bradzacher** [microsoft/TypeScript#53010]
# Suggestion

## 🔍 Search Terms

List of keywords you searched for before creating this issue. Write them down here so that others can find this suggestion more easily and help provide feedback.

performance, parser, node location, loc

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

Currently TS does not compute or store a node's `loc` during a parse.
For TS itself this is fine because it's rarer that TS needs this information (I don't know for sure, but I'd say it only needs it when reporting errors?).
For @typescript-eslint, however, we need to compute the location because ESLint uses it for various things.

The issue is that computing the loc after the parse cycle means that TS has to do a binary search on the line/range table to compute the line/col of each of the start/end - which is obviously not super cheap to do.

As an alternative approach I tried changing ts-eslint's parser to lazily compute the locaction (https://github.com/typescript-eslint/typescript-eslint/pull/6542). Sadly it was a net-neutral change because (I believe) the cost of defining a getter/setter is much more expensive than a property.
Unfortunately based on the way the ESLint's AST is designed we can't introduce a `getLoc` function, and can only defer this by using accessors.

Here is a cpu profile to help illustrate the cost - [NON_LAZY_1.cpuprofile.zip](https://github.com/microsoft/TypeScript/files/10845401/NON_LAZY_1.cpuprofile.zip). If you load this into https://speedscope.app you can use the search to find `SourceFileObject.getLineAndCharacterOfPosition`. It shows that the total time spent in `SourceFileObject.getLineAndCharacterOfPosition` for the parse was ~260ms (0.61% of the type-aware lint run).

---

You people know TS's parser better than I - so you're more aware of whether or not it's feasible to compute the loc up front without a binary search or not. If it's not - then we can just close this out as non-actionable.

**Comment by RyanCavanaugh** [comment]
Just to avoid any possible confusion, what is `loc` in your terminology? Line/Column?

**Comment by bradzacher** [comment]
@RyanCavanaugh correct - by loc I mean the line/column for the start and end of a node.

**Comment by RyanCavanaugh** [comment]
I haven't profiled/implemented it, but if you need line/col info with high chattiness then I would probably take the approach of, for each file, do a linear scan of the source text and build a position -> line/col lookup table (really just two parallel arrays, to avoid allocating a huge number of objects), tack that onto the SourceFile, then put a single getter onto the Node prototype which reads out of its parent SourceFile's lookup tables. That's `O(n)` all-up instead of `O(n log n)` without much memory overhead.

At the risk of summoning cursed knowledge, [not everyone agrees on what constitutes a line](https://github.com/microsoft/language-server-protocol/issues/80), so having this approach on your end means you're able to choose the correct definition of "line" for the given consumer.

**Comment by bradzacher** [comment]
Isn't your suggestion more or less what TS does?
https://github.com/microsoft/TypeScript/blob/c97ddac60d72a4ffe8a77235c2bb2abcb63fbed6/src/compiler/scanner.ts#L475-L489

We call `SourceFile.getLineAndCharacterOfPosition`, which eventually calls the above function, and passes in the line starts computed (and cached) by this function:
https://github.com/microsoft/TypeScript/blob/c97ddac60d72a4ffe8a77235c2bb2abcb63fbed6/src/compiler/scanner.ts#L391-L418

**Comment by RyanCavanaugh** [comment]
No. TS computes a space-efficient, slower-lookup line-to-offset map like
```
0 -> 0
1 -> 5
2 -> 10
3 -> 15
```
What I'm saying is, compute a less-space-efficient, faster-lookup map like
```
// offset to column / line
0 -> 0 0
1 -> 1 0
2 -> 2 0
3 -> 3 0
4 -> 4 0
5 -> 5 0
6 -> 0 1
7 -> 0 2
...
```
That way there's no binary search needed; given an offset you directly get the line/col values.

**Comment by bradzacher** [comment]
Ah, I see. The issue would be that we *probably* want the same space savings as TS because we're keeping the `SourceFile` around for every file that gets parsed.
So if we computed a map like that we'd end up with the range/loc for every node as well as an indefinitely sized map for the lookup.

Which is why it'd be great if we could eagerly compute the loc during the parse as it would save us computing or storing an intermediate map to lookup after-the-fact.
